x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardChangesAction.java

```diff
@@ -274,6 +287,15 @@ protected Response newResponse() {
                     break;
                 }
             }
+        } catch (IllegalStateException e) {
+            // TODO: handle peek reads better.
```

Comment:
I know this is not cool. I'm still thinking about this.

Reply:
So I changed the size semantics, allowing to return less documents then was requested (as long as there are no gaps). But then I ran into a problem that when the shard follow task requested a specific range of ops it knew to be there, a replica shard copy may not have it and thus less ops were returned. This was before I added the logic to deal with less operations being returned because `max_translog_bytes` limit met and I realize now that this problem should be solved in the same way as the `max translog bytes limit` scenario. So I can safely change the maximum size semantics in the shard changes api.

Reply:
I dont' get it. can you point us to the place where the ISE is coming from? Is it due toe source not being available?

Reply:
This is gone now. I did make a change to `LuceneChangesSnapshot`'s `rangeCheck(...)` method to not be strict about not returning up to `toSeqNo`.